### ISSUE TITLE
`avocado run` startup time is very long and nrunner overhead is very large

### ISSUE THREAD

**Issue by verdog** [avocado-framework/avocado#5495]
**Describe the bug**
I am using avocado in an environment where utilizing nrunner's parallel tests isn't possible (each test uses all of a certain resource available on the system)

With a test suite of only 40 tests, avocado takes over a minute to start executing them, and takes over a second to start the next one after one has finished.

Demo:
This gif starts right after I press enter to start avocado.
![demo](https://user-images.githubusercontent.com/25630273/197263444-d0fb6442-217a-4e96-855c-1bf59913cfae.gif)

**Steps to reproduce**
```
import avocado

class test_1(avocado.Test):
    def test(self):
        pass

class test_2(avocado.Test):
    def test(self):
        pass

class test_3(avocado.Test):
    def test(self):
        pass

class test_4(avocado.Test):
    def test(self):
        pass

class test_5(avocado.Test):
    def test(self):
        pass

class test_6(avocado.Test):
    def test(self):
        pass

class test_7(avocado.Test):
    def test(self):
        pass

class test_8(avocado.Test):
    def test(self):
        pass

class test_9(avocado.Test):
    def test(self):
        pass

class test_10(avocado.Test):
    def test(self):
        pass

class test_11(avocado.Test):
    def test(self):
        pass

class test_12(avocado.Test):
    def test(self):
        pass

class test_13(avocado.Test):
    def test(self):
        pass

class test_14(avocado.Test):
    def test(self):
        pass

class test_15(avocado.Test):
    def test(self):
        pass

class test_16(avocado.Test):
    def test(self):
        pass

class test_17(avocado.Test):
    def test(self):
        pass

class test_18(avocado.Test):
    def test(self):
        pass

class test_19(avocado.Test):
    def test(self):
        pass

class test_20(avocado.Test):
    def test(self):
        pass

class test_21(avocado.Test):
    def test(self):
        pass

class test_22(avocado.Test):
    def test(self):
        pass

class test_23(avocado.Test):
    def test(self):
        pass

class test_24(avocado.Test):
    def test(self):
        pass

class test_25(avocado.Test):
    def test(self):
        pass

class test_26(avocado.Test):
    def test(self):
        pass

class test_27(avocado.Test):
    def test(self):
        pass

class test_28(avocado.Test):
    def test(self):
        pass

class test_29(avocado.Test):
    def test(self):
        pass

class test_30(avocado.Test):
    def test(self):
        pass

class test_31(avocado.Test):
    def test(self):
        pass

class test_32(avocado.Test):
    def test(self):
        pass

class test_33(avocado.Test):
    def test(self):
        pass

class test_34(avocado.Test):
    def test(self):
        pass

class test_35(avocado.Test):
    def test(self):
        pass

class test_36(avocado.Test):
    def test(self):
        pass

class test_37(avocado.Test):
    def test(self):
        pass

class test_38(avocado.Test):
    def test(self):
        pass

class test_39(avocado.Test):
    def test(self):
        pass

class test_40(avocado.Test):
    def test(self):
        pass
```
```
avocado run ./tests.py --nrunner-max-parallel-tasks=1
```

Although this happens with any number of parallel tasks.

**Expected behavior**
Startup and test running shouldn't take that long.

**System information (please complete the following information):**
 - OS: `lsb_release -a`
 ```
> lsb_release -a
LSB Version:    n/a
Distributor ID: SUSE
Description:    SUSE Linux Enterprise Server 15 SP3
Release:        15.3
Codename:       n/a
```
 - Avocado version: `avocado -v`
```
> avocado -v
Avocado 98.0
```
 - Avocado installation method: pip, rpm, github ?
pip inside a virtualenv.
- CPU info
64 core 2Ghz server

**Comment by clebergnu** [comment]
Hi @verdog ,

Thanks for reporting this issue.  I'm afraid it may even be two issues (but somewhat related).

Anyway, if you are not aware, Avocado performs a syntax analysis on Python code, while looking for tests.  The reason is that it avoids executing test code on your development environment at `avocado list` time (and also at `avocado run` time if you use, say, the podman spawner).

That code is somewhat complex, and it looks like having many different test class, triggers the issue you found.  On my system, running almost the same command (I only added `/usr/bin/time` as a prefix) results in:

```
RESULTS    : PASS 40 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2022-10-21T15.53-499e59c/results.html
JOB TIME   : 62.43 s
53.66user 5.61system 1:03.57elapsed 93%CPU (0avgtext+0avgdata 53296maxresident)k
0inputs+5584outputs (0major+834182minor)pagefaults 0swaps
```

If I have a test file with one class, but 40 test methods instead, it results in:

```
RESULTS    : PASS 40 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2022-10-21T15.58-9848aeb/results.html
JOB TIME   : 40.07 s
33.03user 3.87system 0:41.12elapsed 89%CPU (0avgtext+0avgdata 52956maxresident)k
0inputs+5632outputs (0major+827027minor)pagefaults 0swaps
```

---

The second issue has to with the level of isolation that nrunner employs indeed.  The current implementation only covers the [STANDALONE_EXECUTABLE](https://avocado-framework.readthedocs.io/en/98.0/api/core/avocado.core.spawners.html?highlight=SpawnMethod#avocado.core.spawners.common.SpawnMethod.STANDALONE_EXECUTABLE) method.  Some time ago, I wrote a [PoC](https://github.com/avocado-framework/avocado/pull/3766/commits/ae57ee78df7f2935e40394cdfc72a34b458cdcef) for the [PYTHON_CLASS](https://avocado-framework.readthedocs.io/en/98.0/api/core/avocado.core.spawners.html?highlight=SpawnMethod#avocado.core.spawners.common.SpawnMethod.PYTHON_CLASS) method.

In short, `PYTHON_CLASS` takes a huge shortcut by reusing the same Python interpreter, but it's limited to running tests on the same environment (even on a different process can be tricky).

For most use cases, I believe that the extra isolation and the freedom to execute tests elsewhere that is possible with the expensive `STANDALONE_EXECUTABLE` method can be offset by running tests in parallel.  It looks like you actually need to run tests serially, so in that case, `PYTHON_CLASS` could be of great help.

Let me know if this makes sense to you, and we can definitely move this issue forward.

**Comment by verdog** [comment]
Hi @clebergnu, thanks for the quick reply.

First, I was able to replicate your speedup by turning the 40 classes into a single class with 40 methods. Thanks for the tip.

About `PYTHON_CLASS`:

You are saying that `PYTHON_CLASS` would have a limitation where one could no longer assume each test is a unique process, right? So things like modifications to globals would persevere between tests? That would actually be helpful instead of a hindrance 😄. We have been doing some workarounds involving the filesystem to try to leave flags between tests (maybe we're just approaching it wrong though!).

Tangential question: in the current state with the 1-class-40-methods form, is each test still a unique process?

**Comment by clebergnu** [comment]
@verdog Just a quick update: I'm doing some more investigation on this issue, and will reply to your questions shortly.

**Comment by clebergnu** [comment]
I've bisected a major performance regression, with the first bad commit being c27203ae19744e859eaaaa467dd58d535333c19d.  At this point, running those 40 classes with empty tests takes (on my system):

```
RESULTS    : PASS 40 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2022-10-28T09.54-12535ca/results.html
JOB TIME   : 51.47 s
```

On the same system, running from the previous commit (59d6f211b64855261b3d561bc2448222c3b0f9d4) results in:

```
RESULTS    : PASS 40 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/cleber/avocado/job-results/job-2022-10-28T09.57-7cdeca1/results.html
JOB TIME   : 30.05 s
```

I'm now looking at what exactly, in that commit, caused the regression.  Marking @richtja as he is the author of that commit and can certainly help to spot what is wrong with it.

**Comment by clebergnu** [comment]
Looks like there's one huge jump in CPU time at the `Runnable` instantiation.

The following command on 59d6f211b64855261b3d561bc2448222c3b0f9d4:

```
python3 -m timeit -s 'from avocado.core.nrunner.runnable import Runnable' -- 'Runnable("noop", "noop")'
500000 loops, best of 5: 608 nsec per loop
```

While under c27203ae19744e859eaaaa467dd58d535333c19d:

```
python3 -m timeit -n 50000 -s 'from avocado.core.nrunner.runnable import Runnable' -- 'Runnable("noop", "noop")'
50000 loops, best of 5: 28.1 usec per loop
```

28.1 usec amount to 28100 nsec, so this is aproximately a 46x increase in cost.  I've found similar situations in other snippets, but these do not seem to be the main culprit.  It seems to have to do with one extra execution of the runner to gather its capabilities.  I'll follow up on that.

**Comment by clebergnu** [comment]
@verdog with pending https://github.com/avocado-framework/avocado/pull/5511 and already merged https://github.com/avocado-framework/avocado/pull/5505, two of the performance problems caught by this issue were solved.

The performance should actually be better now than back in 59d6f211b64855261b3d561bc2448222c3b0f9d4.

---

BTW, if anyone is curious about the difference in performance between having 40 classes and 40 methods.  It was caused because of the PYTHONPATH environment variable growing on every execution.  Because Avocado's safeloader mechanism takes into account the locations where a Python module may be located at, it added considerable more running time to *each class*.  Once one class is zeroed in on, then the heuristics on the methods are not affected.

**Comment by verdog** [comment]
Thanks for your work on this!

**Comment by clebergnu** [comment]
> Thanks for your work on this!

Thanks for reporting the issue! It prevented a release with a badly performing Avocado.

I'm closing this issue.  If you find other issues, please let us know.
